### PR TITLE
Projected volume plugin

### PR DESCRIPTION
**Pull request by jpeeler**
This is a WIP volume driver implementation as noted in the commit for https://github.com/kubernetes/kubernetes/pull/35313.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37237)

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot ok to test

File: test/e2e/common/projected.go [hunk 1]
```diff
@@ -0,0 +1,1518 @@
...
+						},
+					},
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+		},
+	}
+
+	return pod
+}
```
Comment (pmorie) [inline test/e2e/common/projected.go]: Add a test case that projects a secret, configmap, and downward API into the same volume

**Review by pmorie — COMMENTED** [review]
Implementation looks ok - overall my biggest feedback is that we should probably not re-use the API types for the other volumes.

File: pkg/api/types.go [hunk 2]
```diff
@@ -724,6 +726,10 @@ type SecretVolumeSource struct {
 	// Name of the secret in the pod's namespace to use.
 	// +optional
 	SecretName string
+	// same as above, just matches configmap naming for consistency
```
Comment (jpeeler) [inline pkg/api/types.go]: I'll go ahead and remove this. The spec does say "name" should be used with a secret, so if "secretName" is not going to work then I suppose a new API object will be required.

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -753,6 +761,88 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
 	return allErrs
 }
 
+func validateProjectionsVolumeSource(projection *api.Projections, fldPath *field.Path) field.ErrorList {
```
Comment (jpeeler) [inline pkg/api/validation/validation.go]: I tried to be consistent with the naming. Just to be clear, you're talking about changing Projections to Projection throughout? https://github.com/kubernetes/community/pull/169/commits/1383c030484ee693df59cf8b40255954016b2f27#diff-c5fe36fe223ccc59602a515f60c5db37R189

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -753,6 +761,88 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
...
+	allErrs := field.ErrorList{}
+	allPaths := sets.String{}
+
+	projectionMode := projection.DefaultMode
+	if projectionMode != nil && (*projectionMode > 0777 || *projectionMode < 0) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("defaultMode"), *projectionMode, volumeModeErrorMsg))
+	}
+
+	for _, source := range projection.Sources {
```
Comment (jpeeler) [inline pkg/api/validation/validation.go]: I've created a new function for it now.

File: pkg/volume/downwardapi/downwardapi.go [hunk 5]
```diff
@@ -203,17 +203,17 @@ func (b *downwardAPIVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
 	return nil
 }
 
-// collectData collects requested downwardAPI in data map.
+// CollectData collects requested downwardAPI in data map.
 // Map's key is the requested name of file to dump
 // Map's value is the (sorted) content of the field to be dumped in the file.
-func (d *downwardAPIVolume) collectData(defaultMode *int32) (map[string]volumeutil.FileProjection, error) {
+func CollectData(items []v1.DownwardAPIVolumeFile, pod *v1.Pod, host volume.VolumeHost, defaultMode *int32) (map[string]volumeutil.FileProjection, error) {
```
Comment (jpeeler) [inline pkg/volume/downwardapi/downwardapi.go]: I've added the comment here and for MakePayload in the secret and configmap drivers.

File: pkg/volume/projected/projected.go [hunk 6]
```diff
@@ -0,0 +1,327 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (jpeeler) [inline pkg/volume/projected/projected.go]: Updated!

File: test/e2e/common/projected.go (hunk 1 above)
Comment (jpeeler) [inline test/e2e/common/projected.go]: I actually added this test in between original submission and your comments. So this is already done.

File: pkg/api/types.go [hunk 7]
```diff
@@ -724,6 +726,7 @@ type SecretVolumeSource struct {
 	// Name of the secret in the pod's namespace to use.
 	// +optional
 	SecretName string
+
```
Comment (pmorie) [inline pkg/api/types.go]: dead line

File: pkg/api/types.go [hunk 8]
```diff
@@ -740,7 +743,25 @@ type SecretVolumeSource struct {
 	// mode, like fsGroup, and the result can be other mode bits set.
 	// +optional
 	DefaultMode *int32
-	// Specify whether the Secret or it's key must be defined
+	// Specify whether the Secret or its key must be defined
+	// +optional
+	Optional *bool
+}
+
+type SecretProjection struct {
```
Comment (pmorie) [inline pkg/api/types.go]: godoc this type

File: pkg/api/types.go [hunk 9]
```diff
@@ -1011,6 +1038,33 @@ type ConfigMapVolumeSource struct {
 	Optional *bool
 }
 
+type ConfigMapProjection struct {
```
Comment (pmorie) [inline pkg/api/types.go]: godoc all the types

File: pkg/api/v1/types.go [hunk 10]
```diff
@@ -944,6 +946,24 @@ const (
 	SecretVolumeSourceDefaultMode int32 = 0644
 )
 
+type SecretProjection struct {
+	// +optional
+	LocalObjectReference `json:"name,omitempty"`
```
Comment (pmorie) [inline pkg/api/v1/types.go]: The tag for this should be: ``json:",inline"``.  Also, this is not an optional field.

File: pkg/api/types.go (hunk 7 above)
Comment (jpeeler) [inline pkg/api/types.go]: removed

File: pkg/api/types.go (hunk 8 above)
Comment (jpeeler) [inline pkg/api/types.go]: done

File: pkg/api/types.go (hunk 9 above)
Comment (jpeeler) [inline pkg/api/types.go]: done

**Comment by pmorie** [comment]
@k8s-bot test this

File: pkg/api/types.go [hunk 11]
```diff
@@ -746,7 +748,31 @@ type SecretVolumeSource struct {
...
+
+// Adapts a secret into a projected volume.
+//
+// The contents of the target Secret's Data field will be presented in a
+// projected volume as files using the keys in the Data field as the file names.
+// Secret projection volumes support ownership management and SELinux relabeling.
+// Note that this is identical to a secret volume source without the default
+// mode.
+type SecretProjection struct {
+	// +optional
```
Comment (pmorie) [inline pkg/api/types.go]: This is not optional

File: pkg/api/v1/types.go [hunk 12]
```diff
@@ -944,6 +946,30 @@ const (
 	SecretVolumeSourceDefaultMode int32 = 0644
 )
 
+// Adapts a secret into a projected volume.
+//
+// The contents of the target Secret's Data field will be presented in a
+// projected volume as files using the keys in the Data field as the file names.
+// Secret projection volumes support ownership management and SELinux relabeling.
```
Comment (pmorie) [inline pkg/api/v1/types.go]: This part of the comment re: secret projection volumes support .... should only be on the ProjectedVolumeSource, not on the projections.

File: pkg/api/v1/types.go [hunk 13]
```diff
@@ -1108,6 +1134,59 @@ const (
 	ConfigMapVolumeSourceDefaultMode int32 = 0644
 )
 
+// Adapts a ConfigMap into a projected volume.
+//
+// The contents of the target ConfigMap's Data field will be presented in a
+// projected volume as files using the keys in the Data field as the file names,
+// unless the items element is populated with specific mappings of keys to paths.
+// ConfigMap volumes support ownership management and SELinux relabeling.
```
Comment (pmorie) [inline pkg/api/v1/types.go]: Same thing re: ownership management.

File: pkg/api/v1/types.go [hunk 14]
```diff
@@ -4095,6 +4174,16 @@ type DownwardAPIVolumeFile struct {
 	Mode *int32 `json:"mode,omitempty" protobuf:"varint,4,opt,name=mode"`
 }
 
+// Represents downward API info for projecting into a projected volume.
+// Downward API volumes support ownership management and SELinux relabeling.
```
Comment (pmorie) [inline pkg/api/v1/types.go]: Strike this line

**Review by pmorie — COMMENTED** [review]
Just a few nits; fix these and we'll merge and possibly fix up some details in follow-ups.

**Comment by pmorie** [comment]
@k8s-bot unit test this issue: #39471

**Comment by rata** [comment]
@jpeeler thanks for tour time! And really quick response. Sorry to bother :)

**Pull request merged**